### PR TITLE
Prevent seeing light through walls

### LINKED ISSUE #13356: Lab - turret doors blue in curses- but not in tiles-version
In the lab it is very easy to dodge the turret rooms in the ncurses version because their doors are in cyan and you see the range radius. 
In the tiles version you "only" see the adjacent walls outside your nightvision.
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/26008936-lab-turret-doors-blue-in-curses-but-not-in-tiles-version?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by jbytheway**
#### Summary
SUMMARY: Interface "Prevent seeing light through walls"

#### Purpose of change
There's a long-standing issue where walls that are lit from one side appear bright from the other side.  This is strange and unintuitive.

I have a proof-of-concept solution, and wanted to share it for feedback before spending the time to polish it.

#### Describe the solution
First some pictures.
Before:
cata-lighting-before
After:
cata-lighting-after
Works with vehicle lights too:
cata-lighting-headlights
The algorithm is fairly simple.  Rather than storing a single illumination value in the light map, when a tile is opaque (like a wall) we store four values: one for each quadrant of 2D space.  Each value represents the illumination from that quadrant.  When computing apparent visibility, we check which other tiles adjacent to the tile in question we can see, and count only those light values corresponding to the quadrants where we can see adjacent tiles.

This works pretty well in practice, the only small annoyance I've observed is that you get a 'looking around corners' kind of effect in some cases.  This can be seen in the second image above: the wall to the left of the northern door appears bright because we can also see the tile to the NE of it, and it is being lit from that direction.  But that's never giving away any information the player couldn't have deduced, so I think it's fine, and it doesn't look too weird.

Fixes #13356.

#### Performance
No formal measurements yet, but it seems fine in "normal" situations.
There are two pieces of code to consider:
* The light casting code.  When optimized it should be only one extra branch+array index per tile here, so that ought to be fine.  Even in debug build I'd expect negligible overhead.
* Calculating lighting levels.  There will be a largish overhead when calculating lighting level for opaque tiles, but the number of opaque tiles it's possible to see at once is much less than the number of non-opaque ones, so I think this should be fine too.

The current memory layout is rather cache-unfriendly.  If performance testing dictates, it may be wise to swap the array dimensions around.

#### Still to do
- [x] Update the tests (including adding tests verifying that walls are lit correctly in this way).
- [x] Update the 3D FOV code.
- [x] Test performance when many light sources exist.

#### Additional context
I'm aware there has been discussion of a new lighting algorithm (#23996).  I think this change is entirely compatible with that new algorithm.

### CODE REVIEW

**Comment by kevingranade** [comment]
This definitely looks like the right way to do it, I've been concerned that it was going to unduly impact performance, but I hadn't thought of only applying the 4 quadrant update to opaque tiles.

I agree this seems to mesh with all the optimizations I have planned for shadowcasting (both the dynamic lighting overhaul and refactoring the shadowcasting algorithm to be breadth-first).

**Comment by jbytheway** [comment]
The tests comparing various algorithms (`shadowcasting_runoff`, etc.) don't pass on master.  Should they?  I tried rolling back to when they were first added in 4353ec9246 to see if they worked then, but that version requires libtap++, which I can't easily install.

**Comment by jbytheway** [comment]
OK, I found a version that does pass (f48f788d4c); bisecting.

**Comment by jbytheway** [comment]
OK, so it broke in  e858c007e7d5f slightly.  It looks like it might have got worse at some point since then, but it's kind of subjective, so I'm not sure.  In any case, I assume I should not be worrying about that test.  (Should have paid more attention to the comment written above it, I suppose...)

File: src/map.h [hunk 1]
```diff
@@ -161,6 +162,60 @@ struct bash_params {
...
+        return values[static_cast<int>( q )];
+    }
+    void fill( float v ) {
+        std::fill( values.begin(), values.end(), v );
+    }
+    float max() const {
+        return *std::max_element( values.begin(), values.end() );
+    }
+
+    friend inline four_quadrants operator*( const four_quadrants &l, const four_quadrants &r ) {
```
Comment (BevapDin) [inline src/map.h]: `inline` is redundant here. Functions defined inside the class definition are automatically `inline`.

Comment (jbytheway) [reply src/map.h]: I thought that was only true of member functions, but it seems you're right; it applies to friend definitions also.  Thanks.

**Comment by kevingranade** [comment]
Sorry about the confusion, most of those tests are execution aids and migration validation rather then ongoing regression tests.

**Comment by jbytheway** [comment]
Tehre's another oddity I'd like to understand.  If we look at the following chunk of code from the visibility algorithm (this is from master, not my branch):
https://github.com/CleverRaven/Cataclysm-DDA/blob/f99a9637ad53d953904ef06b48e785f1106351af/src/map.cpp#L5458-L5471
we see that it's choosing from a sequence of decreasing light levels.

There's one strangeness which is already called out in a comment in this code.  It can never return `LL_DARK`.

However, there is more.  Look at the definitions of the constants:
https://github.com/CleverRaven/Cataclysm-DDA/blob/f99a9637ad53d953904ef06b48e785f1106351af/src/lightmap.h#L6
https://github.com/CleverRaven/Cataclysm-DDA/blob/f99a9637ad53d953904ef06b48e785f1106351af/src/lightmap.h#L15
`LIGHT_SOURCE_BRIGHT` and `LIGHT_AMBIENT_LIT` are the same, so this function can also never return `LL_LIT`.  Is that intentional?

**Comment by jbytheway** [comment]
Tests written.  Some bugs found and fixed.  I believe it's functionally complete now.  Still want to do some performance testing.

**Comment by jbytheway** [comment]
(Rebased to avoid build error about mismatched Makefile)

I think this is now ready.  Of course, testing by others would be welcome.

Performance testing shows very little difference between the light casting algorithms' performance (up to ~1%):

```
castLight on four_quadrants (denominator 10) executed 1000000 times in 655537 microseconds.
castLight on floats (denominator 10) executed 1000000 times in 652451 microseconds.

castLight on four_quadrants (denominator 100) executed 1000000 times in 652528 microseconds.
castLight on floats (denominator 100) executed 1000000 times in 651208 microseconds.

castLight on four_quadrants (denominator 10) executed 1000000 times in 857010 microseconds.
castLight on floats (denominator 10) executed 1000000 times in 848121 microseconds.

castLight on four_quadrants (denominator 100) executed 1000000 times in 651661 microseconds.
castLight on floats (denominator 100) executed 1000000 times in 655883 microseconds.

castLight on four_quadrants (denominator 10) executed 1000000 times in 651561 microseconds.
castLight on floats (denominator 10) executed 1000000 times in 649798 microseconds.

castLight on four_quadrants (denominator 100) executed 1000000 times in 656125 microseconds.
castLight on floats (denominator 100) executed 1000000 times in 651559 microseconds.

castLight on four_quadrants (denominator 10) executed 1000000 times in 855901 microseconds.
castLight on floats (denominator 10) executed 1000000 times in 846490 microseconds.

castLight on four_quadrants (denominator 100) executed 1000000 times in 651638 microseconds.
castLight on floats (denominator 100) executed 1000000 times in 652841 microseconds.

castLight on four_quadrants (denominator 10) executed 1000000 times in 820022 microseconds.
castLight on floats (denominator 10) executed 1000000 times in 816416 microseconds.

castLight on four_quadrants (denominator 100) executed 1000000 times in 651293 microseconds.
castLight on floats (denominator 100) executed 1000000 times in 652072 microseconds.
```

As anecdotal evidence, I also tried burning down some buildings and the performance seems fine.

I used perf while doing so; it shows that `generate_lightmap` is a large part of the cost in such a situation.  ~35% of samples are in `map::build_map_cache` and 30% in `map::generate_lightmap`.  I think that's pretty much the same as on master (it's hard to construct a repeatable experiment for comparison).

My other reason for running perf was to verify that `map::update_visibility_cache` was not suddenly very expensive.  Indeed it is not, clocking in at ~2% of samples.

**Comment by jbytheway** [comment]
Got a slightly concerning crash on one of the mingw builds on Travis.  Not sure how to debug that.  Is it possible to get the actual binary that was built there so I can interpret the backtrace?

**Comment by kevingranade** [comment]
The container is destroyed, you could possibly stick a command in the travis.yml file to push it somewhere, other than that youd have to try and recreate the docker image travis uses.

**Comment by jbytheway** [comment]
OK, after some investigation it seems my test refactoring had exposed some flakiness in other tests.  All the failures I've observed should be fixed now.  The current test failures are the same vehicle power ones that are failing everywhere.

**Comment by jbytheway** [comment]
I've rebased and tidied up these changes after pulling out #26701 and #26728.  I think it's as clean as it's going to get and good for review.

File: src/lightmap.cpp [hunk 2]
```diff
@@ -447,6 +465,115 @@ float map::light_transparency( const tripoint &p ) const
...
+
+    const bool p_opaque = is_opaque( p.x, p.y );
+    float apparent_light;
+
+    if( p_opaque && vis > 0 ) {
+        // This is the complicated case.  We want to check which quadrants the
+        // player can see the tile from, and only count light values from those
+        // quadrants.
+        using offset_and_quadrants = std::pair<point, std::array<quadrant, 2>>;
+        static constexpr std::array<offset_and_quadrants, 8> adjacent_offsets = {{
```
Comment (ZhilkinSerg) [inline src/lightmap.cpp]: Apparently OSX crosscompile was broken:

```
09:02:18 src/lightmap.cpp:486:62: error: constexpr variable 'adjacent_offsets' must be initialized by a constant expression
09:02:18         static constexpr std::array<offset_and_quadrants, 8> adjacent_offsets = {{
09:02:18                                                              ^                  ~~
09:02:18 src/lightmap.cpp:487:17: note: non-constexpr constructor 'pair' cannot be used in a constant expression
09:02:18                 { { 0, 1 }, {{ quadrant::SE, quadrant::SW }} },
09:02:18                 ^
09:02:18 /usr/local/osxcross/osxcross/target/bin/../SDK/MacOSX10.11.sdk/usr/include/c++/v1/utility:267:5: note: declared here
09:02:18     pair(const _T1& __x, const _T2& __y)
09:02:18     ^
```

Comment (jbytheway) [reply src/lightmap.cpp]: Fixed in #26786.

**Pull request merged**